PCF-ADR-0007-operator-redesign.md

```diff
@@ -0,0 +1,189 @@
+
+Currently each component has its own way of sourcing configuration:
+* CLI options for MCAD
+* CLI options, and a ConfigMap for InstaScale
+* CLI options, and CRDs for the CodeFlare operator
+
+We propose to model the configuration of each component as part of their public API.
+
+MCAD would declare its configuration, e.g.:
+
+```go
+package config
+
+struct MCADConfiguration {
+	// Number of seconds a job will go away for, if it can not be scheduled.
+	// Default is 20.
+	BackoffTime int `json:"backoffTime,omitempty"`
+	// Head of line job will not be bumped away for at least HeadOfLineHoldingTime seconds
+	// by higher priority jobs.
+	// Default setting to 0 disables this mechanism.
+	HeadOfLineHoldingTime int `json:"headOfLineHoldingTime,omitempty"`
+
+	// Remaining configuration...
+}
+```
+
+InstaScale would also declare its configuration, e.g.:
+
+```go
+package config
+
+struct InstaScaleConfiguration {
+	// Maximum of cluster nodes that can be scaled out.
+	// Defaults to 3.
+	MaxScaleoutAllowed int `json:"maxScaleoutAllowed,omitempty"`
+
+	// Remaining configuration...
+}
+```
+
+Finally, the CodeFlare operator would embed MCAD and InstaScale configurations, along with the generic configuration taken from [component-base](https://github.com/kubernetes/component-base), e.g.:
+
+```go
+package config
+
+import (
+	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
+	mcad "github.com/project-codeflare/multi-cluster-app-dispatcher/pkg/config"
+	instascale "github.com/project-codeflare/instascale/pkg/config"
+)
+
+struct CodeFlareOperatorConfiguration {
+	// The MCAD controller configuration
+	MCAD *mcad.MCADConfiguration `json:"mcad,omitempty"`
+
+	// The InstaScale controller configuration
+	InstaScale *instascale.InstaScaleConfiguration `json:"instascale,omitempty"`
+	// ClientConnection configures the connection to the cluster
+	ClientConnection *configv1alpha1.ClientConnectionConfiguration `json:"clientConnection,omitempty"`
+
+	// LeaderElection is the LeaderElection config to be used when configuring
+	// the manager.Manager leader election
+	LeaderElection *configv1alpha1.LeaderElectionConfiguration `json:"leaderElection,omitempty"`
+
+	// Remaining configuration...
+}
+```
+
+The later `struct` would be sourced from a ConfigMap, mounted into the operator deployment container.
+To follows the Operator SDK best practice, which states an operator [should always be able to deploy and come up without user input](https://sdk.operatorframework.io/docs/best-practices/best-practices/#summary-1), that ConfigMap would be created when the operator starts, if it doesn't exist already, and initialized with sensible defaults.
+
+We might want to package that ConfigMap as an [additional resource](https://olm.operatorframework.io/docs/tasks/creating-operator-manifests/#packaging-additional-objects-alongside-an-operator) in the OLM bundle.
```

Comment:
How would this work if the user supplies their own configuration? Would installing the OLM bundle overwrite the existing values?

Reply:
Ah that's a very good point. I'm pretty sure OLM would override the user provided configuration.

I've removed that paragraph in 6f24d9704dbcdc8f8741e36acc71c09afa54f409.